python/cudf/cudf/core/column/column.py

```diff
@@ -1481,6 +1481,12 @@ def as_column(arbitrary, nan_as_null=None, dtype=None, length=None):
 
         if dtype is not None:
             arbitrary = arbitrary.astype(dtype)
+        elif (
+            shape
+            and shape[0]
+            and hasattr(arbitrary[0], "__cuda_array_interface__")
+        ):
+            arbitrary = arbitrary.astype(arbitrary[0].dtype)
```

Comment:
```suggestion
        if (
            shape
            and shape[0]
            and hasattr(arbitrary[0], "__cuda_array_interface__")
        ):
            return as_column(
                cupy.asarray(arbitrary, dtype=arbitrary[0].dtype),
                nan_as_null=nan_as_null,
                dtype=dtype,
                length=length,
            )
```


1. This will actually lead to D->H copying of data in `arbitrary` and we'd want to avoid it. You can construct a cupy array out of arbitrary.
```python
>>> x = np.array([cp.array(0), cp.array(1)])
>>> x
array([array(0), array(1)], dtype=object)
>>> type(x)
<class 'numpy.ndarray'>
>>> x.astype(x[0].dtype)
array([0, 1])
>>> type(x.astype(x[0].dtype))
<class 'numpy.ndarray'>
>>> cp.asarray(x, dtype=x[0].dtype)
array([0, 1])
>>> type(cp.asarray(x, dtype=x[0].dtype))
<class 'cupy.core.core.ndarray'>
>>> 
```
2. Any specific reason that we only want to do this check when `dtype` is None? If `dtype` is not None, we'll again end up creating a D->H copy too. Instead, we should handle this entire `if` case always after line number 1478.

Reply:
> This will actually lead to D->H copying of data in arbitrary and we'd want to avoid it. You can construct a cupy array out of arbitrary and let the column constructor handle the rest.

Sounds good - Thanks for the suggestion

> Any specific reason that we only want to do this check when dtype is None? If dtype is not None, we'll again end up creating a D->H copy too. Instead, we should handle this entire if case always after line number 1478.

I was going for a quick minimal change here, but I agree we should improve existing code as well.  Thanks for taking the time to think through this!